### PR TITLE
Symlink header folders by script build phase

### LINKED ISSUE #5313: Unable to import OS X frameworks using private headers and header_mappings_dir
## What did you do?

Integrated Realm in an OS X target.

## What did you expected to happen?

The integrating target would build.

## What happened instead?

The generated pod target for Realm itself builds, but it can't be imported. Clang complains about missing headers from the module map:

```
Realm.framework/Modules/module.modulemap:8:16: error: header 'RLMAccessor.h' not found
        header "RLMAccessor.h"
               ^

DataImporter.swift:20:8: error: could not build Objective-C module 'Realm'
import Realm
       ^
```

Inspecting the build products and comparing them with the binaries we build ourselves from our Xcode project, it turns out, that the `PrivateHeaders` symlink in the root directory of the framework bundle is missing. While the actual `PrivateHeaders` dir exists within `Versions/A`. My suspicion is that the custom copy files build phase, which ensures to retain the header layout of the `header_mappings_dir` creates the `PrivateHeaders` directory in a way Xcode doesn't expect, preventing the automatic creation of the symlink.

So this seems to be an instance of a more general error, which is only relevant for podspecs using `header_mappings_dir`, a custom module map and private headers, which are explicitly declared in the module map.

## CocoaPods Environment

```
   CocoaPods : 1.0.0
       Xcode : 7.3 (7D175)
```

## Project that demonstrates the issue

https://github.com/realm/realm-cocoa-converter/tree/to-0.102.0-update

### PR DESCRIPTION
**Pull request by mrackwitz**
Symlink the header folders in the framework bundle's root directory by script build phase if header_mappings_dir is used with frameworks.
Fixes #5313.

### Involved Build Settings

```shell
# Inherited build settings
PUBLIC_HEADERS_FOLDER_PATH  = $(CONTENTS_FOLDER_PATH)/PrivateHeaders
PRIVATE_HEADERS_FOLDER_PATH = $(CONTENTS_FOLDER_PATH)/Headers

# Derived build settings
## Relies on assumptions, because actual definitions aren't documented
WRAPPER_NAME                = $(PRODUCT_NAME)$(WRAPPER_EXTENSION)
CONTENTS_FOLDER_PATH        = $(WRAPPER_NAME)/Versions/$(FRAMEWORK_VERSION)
```

### CODE REVIEW

**Comment by segiddins** [comment]
CHANGELOG?

**Comment by mrackwitz** [comment]
@segiddins @neonichu: Could you review the [contents of the shell script build phase](https://github.com/CocoaPods/CocoaPods/pull/5352/files#diff-b330fa5e79c6b5e63c36c4ff6b333cdfR234)?

File: lib/cocoapods/installer/target_installer/pod_target_installer.rb [hunk 1]
```diff
@@ -216,6 +217,26 @@ def create_xcconfig_file
...
+      #
+      # @return [void]
+      #
+      def create_build_phase_to_symlink_header_folders
+        return unless target.platform.name == :osx && header_mappings_dir
+
+        build_phase = native_target.new_shell_script_build_phase('Create Symlinks to Header Folders')
+        build_phase.shell_script = <<-eos.strip_heredoc
+          cd "$CONFIGURATION_BUILD_DIR/$WRAPPER_NAME"
+          ln -fs ${PUBLIC_HEADERS_FOLDER_PATH\#$WRAPPER_NAME/} ${PUBLIC_HEADERS_FOLDER_PATH\#\$CONTENTS_FOLDER_PATH/}
```
Comment (segiddins) [inline lib/cocoapods/installer/target_installer/pod_target_installer.rb]: what does the `#` do here?

**Comment by segiddins** [comment]
Looks good to me!

**Comment by neonichu** [comment]
LGTM as well :+1:

**Comment by mrackwitz** [comment]
FYI: I just came up with a whole different idea. I thought maybe, if we would move the custom copy file build phases before the built-in copy header build phase, Xcode might be able to figure out that there are public and private headers, but no luck. It only cares about the files, it created itself. Maybe we should create a radar about that.

**Pull request merged**